tensorflow/c/experimental/gradients/nn_grad_test.cc

```diff
@@ -0,0 +1,137 @@
+#include "tensorflow/c/experimental/gradients/grad_test_helper.h"
+#include "tensorflow/c/experimental/gradients/tape/tape_context.h"
+#include "tensorflow/core/platform/test.h"
+
+namespace tensorflow {
+namespace gradients {
+namespace internal {
+namespace {
+
+using tensorflow::TF_StatusPtr;
+using tracing::TracingOperation;
+
+Status BiasAddModel(AbstractContext* ctx,
+                    absl::Span<AbstractTensorHandle* const> inputs,
+                    absl::Span<AbstractTensorHandle*> outputs,
+                    const GradientRegistry& registry) {
+  return ops::BiasAdd(ctx, inputs, outputs, "BiasAdd");
+}
+
+Status BiasAddGradModel(AbstractContext* ctx,
+                        absl::Span<AbstractTensorHandle* const> inputs,
+                        absl::Span<AbstractTensorHandle*> outputs,
+                        const GradientRegistry& registry) {
+  TapeVSpace vspace(ctx);
+  auto tape = new Tape(/*persistent=*/false);
+  tape->Watch(ToId(inputs[0]));  // Watch A.
+  tape->Watch(ToId(inputs[1]));  // Watch Bias.
+  std::vector<AbstractTensorHandle*> temp_outputs(1);
+  AbstractContextPtr tape_ctx(new TapeContext(ctx, tape, registry));
+  TF_RETURN_IF_ERROR(ops::BiasAdd(tape_ctx.get(), inputs,
+                                  absl::MakeSpan(temp_outputs), "BiasAddGrad"));
+  std::unordered_map<tensorflow::int64, TapeTensor>
+      source_tensors_that_are_targets;
+
+  std::vector<AbstractTensorHandle*> out_grads;
+  TF_RETURN_IF_ERROR(tape->ComputeGradient(
+      vspace, /*target_tensor_ids=*/{ToId(temp_outputs[0])},
+      /*source_tensor_ids=*/{ToId(inputs[0]), ToId(inputs[1])},
+      source_tensors_that_are_targets,
+      /*output_gradients=*/{}, &out_grads,
+      /*build_default_zeros_grads=*/false));
+  for (auto temp_output : temp_outputs) {
+    temp_output->Unref();
+  }
+  outputs[0] = out_grads[0];
+  outputs[1] = out_grads[1];
+  delete tape;
+  return Status::OK();
+}
+
+Status RegisterGradients(GradientRegistry* registry) {
+  TF_RETURN_IF_ERROR(registry->Register("BiasAdd", BiasAddRegisterer));
+  return Status::OK();
+}
+
+class CppGradients
+    : public ::testing::TestWithParam<std::tuple<const char*, bool, bool>> {
+ protected:
+  void SetUp() override {
+    TF_StatusPtr status(TF_NewStatus());
+    TF_SetTracingImplementation(std::get<0>(GetParam()), status.get());
+    Status s = StatusFromTF_Status(status.get());
+    CHECK_EQ(errors::OK, s.code()) << s.error_message();
+
+    {
+      AbstractContext* ctx_raw = nullptr;
+      Status s =
+          BuildImmediateExecutionContext(std::get<1>(GetParam()), &ctx_raw);
+      ASSERT_EQ(errors::OK, s.code()) << s.error_message();
+      ctx_.reset(ctx_raw);
+    }
+
+    s = RegisterGradients(&registry_);
+    ASSERT_EQ(errors::OK, s.code()) << s.error_message();
+  }
+
+  GradientRegistry registry_;
+  AbstractContextPtr ctx_;
+};
+
+TEST_P(CppGradients, TestBiasAddGrad) {
+  if (std::get<0>(GetParam()) == "mlir" && !std::get<2>(GetParam())) {
```

Comment:
`std::get<0>(GetParam()) == "mlir"` also complains. Change to `strcmp(std::get<0>(GetParam()), "mlir") == 0 `

Ideally we should use the pattern in unified_api_test: https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/c/eager/unified_api_test.cc;l=55-58;drc=0dd94c4ad3046f4e383b371c60f272e72f9d2525

Reply:
Done